### ISSUE TITLE
vim.treesitter.foldexpr: attempt to index field 'foldupdate_range' (a nil value)

### ISSUE THREAD

**Issue by TheLeoP** [neovim/neovim#28516]
### Problem

Using `vim.opt.foldexpr = "v:lua.vim.treesitter.foldexpr()"` while triggering some kind of autocompletion on insert mode and then going to normal mode without closing the autocompletion menu causes the following error:

```lua
Se ha detectado un error al procesar InsertLeave Auto-órdenes para "<buffer=1>":
Error executing lua callback: ...s\Neovim\share\nvim\runtime/lua/vim/treesitter/_fold.lua:285: attempt to index field 'foldupdate_range' (a nil value)
stack traceback:
        ...s\Neovim\share\nvim\runtime/lua/vim/treesitter/_fold.lua:285: in function 'do_foldupdate'
        ...s\Neovim\share\nvim\runtime/lua/vim/treesitter/_fold.lua:274: in function <...s\Neovim\share\nvim\runtime/lua/vim/treesitter/_fold.lua:273>
```

### Steps to reproduce

1. `nvim -u minimal.lua minimal.lua`
2. `G` go to last line, `o` insert new line under current line and enter insert mode
3. `local function something() end`
4. Without closing the `pummenu`, `<esc>` go to normal mode
5. The error is shown

minimal.lua:
```lua
function ExampleOmnifunc(find_start, base)
	if find_start == 1 then
		return #vim.api.nvim_get_current_line()
	elseif find_start == 0 then
		return { "end" }
	end
end

vim.opt.omnifunc = "v:lua.ExampleOmnifunc"

vim.opt.foldenable = true
vim.opt.completeopt = { "menuone", "noselect", "noinsert" }
vim.opt.foldcolumn = "1"
vim.o.foldlevelstart = 99
vim.o.foldmethod = "expr"
vim.opt.foldexpr = "v:lua.vim.treesitter.foldexpr()"

local triggers = { "d" }
vim.api.nvim_create_autocmd("InsertCharPre", {
	buffer = vim.api.nvim_get_current_buf(),
	callback = function()
		if vim.fn.pumvisible() == 1 or vim.fn.state("m") == "m" then
			return
		end
		local char = vim.v.char
		if vim.list_contains(triggers, char) then
			local key = vim.keycode("<C-x><C-o>")
			vim.api.nvim_feedkeys(key, "m", false)
		end
	end,
})
```

### Expected behavior

There should be no errors and folds should work as expected

### Neovim version (nvim -v)

NVIM v0.10.0-dev-2999+g38b9c322c

### Vim (not Nvim) behaves the same?

Doesn't apply

### Operating system/version

Windows 11 Pro 23H2

### Terminal name/version

Doesn't apply

### $TERM environment variable

Doesn't apply

### Installation

nvim-win64.msi from the github releases page

**Comment by justinmk** [comment]
https://github.com/neovim/neovim/blob/47dbda97d2f40729733b1c0d1d13d914065af23c/runtime/lua/vim/treesitter/_fold.lua#L285

in all other places the logic checks `if self.foldupdate_range` first.

**Comment by tomtomjhj** [comment]
Can redproduce. My intention with `do_foldupdate` not checking `if self.foldupdate_range` was that this function is called after `foldupdate_range` set by `foldupdate`. But it seems there is an interesting interaction between `foldupdate`'s `InsertLeave` autocmd and this issues's `InsertChaPre`. I'll investigate.

**Comment by TheLeoP** [comment]
If

https://github.com/neovim/neovim/blob/47dbda97d2f40729733b1c0d1d13d914065af23c/runtime/lua/vim/treesitter/_fold.lua#L261

is changed to

```lua
  if api.nvim_get_mode().mode:match('^i') then
```

the error goes away

**Comment by TheLeoP** [comment]
From `:help mode()`

> 		   i	    Insert
>		   ic	    Insert mode completion |compl-generic|
>		   ix	    Insert mode |i_CTRL-X| completion

**Comment by tomtomjhj** [comment]
Can confirm that your fix works. can you make a PR?

**Comment by kapral18** [comment]
this is still happening

**Comment by kapral18** [comment]
this is still happening in nvim 0.10 https://github.com/LazyVim/LazyVim/issues/3569

**Comment by s1n7ax** [comment]
@kapral18 Can confirm this is still happening in 0.11. I'm also using LazyVim and this occurs in Typescript, Lua, Java files.

This occurs almost 70% time when I organize imports in Typescript files or use `vim.lsp.buf.code_action()` to import something. And then sometimes when I use treesitter.

**Comment by tomtomjhj** [comment]
#29581 fixes this

**Comment by kapral18** [comment]
@s1n7ax this indeed has to do with ctrl-c, I remapped it to esc and errors stopped, so @tomtomjhj  PR is on point

**Comment by nfwyst** [comment]
this issue still exists in 0.10.3

**Comment by tomtomjhj** [comment]
make backport pr: https://github.com/neovim/neovim/pull/32102

It will be included in 0.10.4 (if released).
